aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -685,6 +685,513 @@ Tensor& matmul_out(Tensor &result, const Tensor & tensor1, const Tensor & tensor
+      1.36783778460411720168e+00,
+      4.98289622525382669416e-01,
+      -6.37898194594723280150e-04
+    },
+    {
+      -1.09676396052962061844e+01,
+      1.68015813878906206114e+00,
+      5.71779846478865511061e-02,
+      -6.98210122488052056106e-03,
+      3.34975017086070470649e-05
+    },
+    {
+      -9.04316832390810593223e-02,
+      -6.76404519071381882256e-02,
+      6.75961301770459654925e-02,
+      2.95552570429315521194e-02,
+      -1.39180257516060693404e-05
+    },
+    {
+      0.,
+      0.,
+      -9.23364619367118555360e-02,
+      -1.69364939002081722752e-02,
+      -1.40086798182036094347e-05
+    }
+  }};
+
+  // gather coefficients `b` from above into a tensor,
+  // and move them to device `device_of(A)`
+  auto bs = at::from_blob(
+    reinterpret_cast<void*>(&b),
+    {num_prods, num_prods},
+    {num_prods, 1},
+    A.dtype()
+  );
+  bs = _pin_and_move_memory_if_cuda_input(bs, A);
+
+  auto As = _allocate_buffer(A, num_prods);
+  _fill_matrix_powers(As, A, num_prods);
+
+  auto Bs = at::native::_compute_linear_combination(As, bs);
+
+  // compute A9
+  Bs.select(0, 3).add_(at::matmul(Bs.select(0, 0), Bs.select(0, 4)));
+  return Bs.select(0, 1).add_(at::matmul(
+    Bs.select(0, 2).add_(Bs.select(0, 3)),
+    Bs.select(0, 3)
+  ));
+}
+
+template <typename scalar_t>
+void compute_T18_scale_square(
+  Tensor& mexp_out,
+  const Tensor& a,
+  const Tensor& norm,
+  scalar_t theta
+) {
+  // Scale
+  const auto s = at::max(
+    at::zeros_like(norm),
+    at::ceil(at::log2(norm / theta))
+  ).unsqueeze(-1).unsqueeze(-1).to(at::kLong);
+  const auto pow2s = at::pow(2, s);
+  const auto a_scaled = a / pow2s;
+
+  // Square
+  auto mexp_scaled = at::native::compute_T18<scalar_t>(a_scaled);
+  auto s_cpu = s.to(at::kCPU);
+  for (int64_t i = 0; i < mexp_scaled.size(0); ++i) {
+    auto s_val = s_cpu.select(0, i).template item<int64_t>();
+    auto mexp = mexp_scaled.select(0, i);
+    for (int64_t p = 0; p < s_val; ++p) {
+      mexp = at::matmul(mexp, mexp);
+    }
+    mexp_out.select(0, i).copy_(mexp);
+  }
+}
+
+template <typename scalar_t>
+Tensor mexp_impl(
+  const Tensor& a,
+  std::array<scalar_t, total_n_degs> thetas,
+  bool compute_highest_degree_approx = false
+) {
+  auto res = at::empty_like(a);
+  const auto norm = operator_1_norm(a);
+  const auto norm_cpu = norm.to(at::kCPU);
+
+  if (!compute_highest_degree_approx) {
+    constexpr std::array<
+      Tensor(*)(const Tensor&),
+      total_n_degs - 1> 
+    compute_Ts = {
+      compute_T1, compute_T2, compute_T4<scalar_t>,
+      compute_T8<scalar_t>, compute_T12<scalar_t>
+    };
+
+    for (int i = 0; i < total_n_degs - 1; ++i) {
+      auto norm_lower_bound = (i == 0) ? static_cast<scalar_t>(-1) : thetas[i - 1];
+      auto norm_upper_bound = thetas[i];
+      // nonzero returns a 2D tensor, hence squeeze(-1) to make it 1D
+      auto idx_curr_norm_interval = (
+        (norm_lower_bound < norm_cpu) * (norm_cpu <= norm_upper_bound)
```

Comment:
It might be clearer and more direct to use logical `&&` here

Reply:
```
../aten/src/ATen/native/LinearAlgebra.cpp:1054:39: error: no match for 'operator&&' (operand types are 'at::Tensor' and 'at::Tensor')             
       auto idx_curr_norm_interval = (                                                                                                            
                                     ~
         (norm_lower_bound < norm_cpu) && (norm_cpu <= norm_upper_bound)
         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
       ).nonzero().squeeze(-1);
       ~

```

Reply:
It might be better to do the norm comparisons on GPU, since there isn't any reason why they can't be accelerated. `nonzero` would then be your syncpoint, but most of the substantive computation still would happen on GPU.

Reply:
In particular, you wouldn't have to move the nonzero indices *back* to GPU; if you computed them there they'd already be where you need them.

Reply:
I tried it, and it showed a non-negligible regression.

Reply:
It is a one single sync vs 7 syncs dilemma, and for CUDA it showed to be more efficient to sync only once.

Reply:
The syncing happens in the very beginning, so the CPU does not have to wait until these heavy matrix products are done.

Reply:
SGTM. Worth mentioning this in the code, esp if you did the counterfactual experiment!

Reply:
OK, will do!